log4j2/src/main/java/com/google/common/flogger/backend/log4j2/Log4j2SimpleLogEvent.java

```diff
@@ -80,54 +106,67 @@ Level getLevel() {
+    //          .getContextApiSingleton()
+    //          .newContext()
+    //          .withMetadata(COUNT_KEY, 23)
+    //          .withTags(Tags.builder().addTag("foo").addTag("baz", "bar").addTag("baz", "bar2").build())
+    //          .install()) {
     //
-    // By using '%X{key}' in the ConversionPattern of an appender this data can be included in the
-    // logs.
+    //      // do business logic that triggers logs
+    // }
     //
-    // We could include this data here by doing 'MDC.getContext()', but we don't want to encourage
-    // people using the log4j specific MDC. Instead this should be supported by a LoggingContext and
-    // usage of Flogger tags.
-    Map<String, String> mdcProperties = Collections.emptyMap();
+    // The statement above will result in adding the following data to the context map:
+    // {count=23, tags=[foo, baz=bar2, baz=bar]}.
+    // By using '%X{key}' or '%X' in the ConversionPattern of an appender the metadata can be included in the
+    // logs.
 
     // The fully qualified class name of the logger instance is normally used to compute the log
     // location (file, class, method, line number) from the stacktrace. Since we already have the
     // log location in hand we don't need this computation. By passing in null as fully qualified
     // class name of the logger instance we ensure that the log location computation is disabled.
     // this is important since the log location computation is very expensive.
     return Log4jLogEvent.newBuilder()
-        .setLoggerName(logger.getName())
-        .setLoggerFqcn(null)
-        .setLevel(level)
-        .setMessage(new SimpleMessage(message))
-        .setThreadName(Thread.currentThread().getName())
-        // Don't use Duration here as (a) it allocates and (b) we can't allow error on overflow.
-        .setTimeMillis(TimeUnit.NANOSECONDS.toMillis(logData.getTimestampNanos()))
-        .setThrown(thrown != null ? Throwables.getRootCause(thrown) : null)
-        .setIncludeLocation(true)
-        .setSource(getLocationInfo())
-        .setContextMap(mdcProperties)
-        .build();
+            .setLoggerName(logger.getName())
+            .setLoggerFqcn(null)
+            .setLevel(level)
+            .setMessage(new SimpleMessage(message))
+            .setThreadName(Thread.currentThread().getName())
+            // Don't use Duration here as (a) it allocates and (b) we can't allow error on overflow.
+            .setTimeMillis(TimeUnit.NANOSECONDS.toMillis(logData.getTimestampNanos()))
+            .setThrown(thrown != null ? Throwables.getRootCause(thrown) : null)
+            .setIncludeLocation(true)
+            .setSource(getLocationInfo())
+            .setContextData(createContextMap())
+            .build();
+  }
+
+  private StringMap createContextMap() {
+    StringMap contextData = ContextDataFactory.createContextData(logData.getMetadata().size());
```

Comment:
I think you want to get the metadata processor first, and then use the size of the "keySet" from it to size the output, and then process the metadata. That's the total number of unique keys (not including clashing labels, but that's super rare).

Reply:
done